resource mysql_user with auth_plugin caching_sha2_password needs to be in hex format

MySQL 8.4 

From the mysql doc (https://dev.mysql.com/doc/refman/8.4/en/show-create-user.html): 

> Password hash values displayed in the IDENTIFIED WITH clause of output from [SHOW CREATE USER](https://dev.mysql.com/doc/refman/8.4/en/show-create-user.html) may contain unprintable characters that have adverse effects on terminal displays and in other environments. Enabling the [print_identified_with_as_hex](https://dev.mysql.com/doc/refman/8.4/en/server-system-variables.html#sysvar_print_identified_with_as_hex) system variable causes [SHOW CREATE USER](https://dev.mysql.com/doc/refman/8.4/en/show-create-user.html) to display such hash values as hexadecimal strings rather than as regular string literals. Hash values that do not contain unprintable characters still display as regular string literals, even with this variable enabled.

Using the caching_sha2_password auth_plugin fails with "Error: failed executing SQL: Error 1827 (HY000): The password hash doesn't have the expected format." To avoid this error, we need to be able to provide the password in binary/hexadecimal string. The catch is when we use the hex string, we don't want to quote it: 

```
CREATE USER for u1@localhost: CREATE USER `u1`@`localhost`
IDENTIFIED WITH 'caching_sha2_password'
AS 0x244124303035240C7745603626313D613C4C10633E0A104B1E14135A544A7871567245614F4872344643546336546F624F6C7861326932752F45622F4F473273597557627139
REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK
PASSWORD HISTORY DEFAULT PASSWORD REUSE INTERVAL DEFAULT
PASSWORD REQUIRE CURRENT DEFAULT
```

How to reproduce the issue:
1) create a user in mysql: 
```
MySQL (mysql)> create user 'testuser'@'localhost' identified by "$%JHYKuu888*)()(";
```
2) get the show create user syntax from mysql: 
```
MySQL (mysql)> show create user 'testuser'@'localhost';
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER for testuser@localhost                                                                                                                                                                                                                                                                     |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE USER `testuser`@`localhost` IDENTIFIED WITH 'caching_sha2_password' AS '$A$005$\n(U*EPgw\ZtdF2fdLzi4eYt7e8pK1GLhI8nH19tUXzwOkqO3zoK7Pz1' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK PASSWORD HISTORY DEFAULT PASSWORD REUSE INTERVAL DEFAULT PASSWORD REQUIRE CURRENT DEFAULT |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
3) drop the user in mysql

4) using terraform , create a variable for the password and copy the password above into variables.tfvar (need to escape the "\\" chars)

5) create a mysql_user resource: 
```
resource "mysql_user" "testuser" {
  user               = "testuser"
  host               = "localhost"
  auth_plugin        = "caching_sha2_password"
  auth_string_hashed = var.auth_hash.testuser
  tls_option         = "SSL"
}
```

6) run terraform apply
```
➜  dev git:(main) ✗ export TF_LOG=ERROR                                                                                                                                                                                                                                                                                            conda:base
➜  dev git:(main) ✗ terraform apply                                                                                                                                                                                                                                                                                                ...
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
Terraform will perform the following actions:
  # module.mysql_resources.mysql_user.testuser will be created
  + resource "mysql_user" "testuser" {
      + auth_plugin          = "caching_sha2_password"
      + auth_string_hashed   = (sensitive value)
      + discard_old_password = false
      + host                 = "localhost"
      + id                   = (known after apply)
      + tls_option           = "SSL"
      + user                 = "testuser"
    }
Plan: 1 to add, 0 to change, 0 to destroy.
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.
  Enter a value: yes
module.mysql_resources.mysql_user.testuser: Creating...
2025-04-16T11:50:16.293-0700 [ERROR] provider.terraform-provider-mysql_v3.0.71: Response contains error diagnostic: diagnostic_detail="" tf_resource_type=mysql_user @caller=github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_severity=ERROR diagnostic_summary="failed executing SQL: Error 1827 (HY000): The password hash doesn't have the expected format." tf_proto_version=5.8 tf_provider_addr=provider tf_req_id=20cd0cc5-5fb2-17df-69bd-2ee08bc1eb9f tf_rpc=ApplyResourceChange timestamp=2025-04-16T11:50:16.293-0700
2025-04-16T11:50:16.301-0700 [ERROR] vertex "module.mysql_resources.mysql_user.testuser" error: failed executing SQL: Error 1827 (HY000): The password hash doesn't have the expected format.
╷
│ Error: failed executing SQL: Error 1827 (HY000): The password hash doesn't have the expected format.
│ 
│   with module.mysql_resources.mysql_user.testuser,
│   on ../../../../modules/mysql_resources/on-premise/ipsk/users.tf line 33, in resource "mysql_user" "testuser":
│   33: resource "mysql_user" "testuser" {
```

--- a/mysql/resource_user.go
+++ b/mysql/resource_user.go
@@ -99,6 +99,7 @@ func resourceUser() *schema.Resource {
 				Type:             schema.TypeString,
 				Optional:         true,
 				Sensitive:        true,
+				StateFunc:        NormalizeHexStringStateFunc,
 				DiffSuppressFunc: SuppressHexStringDiff,
 				ConflictsWith:    []string{"plaintext_password", "password", "auth_string_hashed"},
 			},
@@ -595,7 +596,6 @@ func SuppressHexStringDiff(k, old, new string, d *schema.ResourceData) bool {
 	if normalizedOld == normalizedNew {
 		return true
 	}
-
 	return false
 }
 
@@ -615,6 +615,15 @@ func validateHexString(hexStr string) error {
 	}
 
 	return nil
+}
+
+func NormalizeHexStringStateFunc(val interface{}) string {
+	if val == nil {
+		return ""
+	}
+
+	hexStr := val.(string)
+	return normalizeHexString(hexStr) // Always store normalized format
 }
 
 // Add this helper function to normalize hex strings

--- a/mysql/resource_user.go
+++ b/mysql/resource_user.go
@@ -99,7 +99,7 @@ func resourceUser() *schema.Resource {
 				Type:             schema.TypeString,
 				Optional:         true,
 				Sensitive:        true,
-				DiffSuppressFunc: NewEmptyStringSuppressFunc,
+				DiffSuppressFunc: SuppressHexStringDiff,
 				ConflictsWith:    []string{"plaintext_password", "password", "auth_string_hashed"},
 			},
 			"tls_option": {
@@ -177,24 +177,23 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 			authStm = fmt.Sprintf("%s AS ?", authStm)
 		}
 	}
-	var hashed_hex string
+	var hashedHex string
 	if v, ok := d.GetOk("auth_string_hex"); ok {
-		hashed_hex = v.(string)
-		if hashed_hex != "" {
+		hashedHex = v.(string)
+		if hashedHex != "" {
 			if hashed != "" {
 				return diag.Errorf("can not specify both auth_string_hashed and auth_string_hex")
 			}
 			if authStm == "" {
 				return diag.Errorf("auth_string_hex is not supported for auth plugin %s", auth)
 			}
-			if strings.HasPrefix(hashed_hex, "0x") || strings.HasPrefix(hashed_hex, "0X") {
-				hashed_hex = hashed_hex[2:]
-			}
+			normalizedHex := normalizeHexString(hashedHex)
+			hexDigits := normalizedHex[2:] // Remove the "0x" prefix for validation
 
-			if err := validateHexString(hashed_hex); err != nil {
+			if err := validateHexString(hexDigits); err != nil {
 				return diag.Errorf("invalid hex string for auth_string_hex: %v", err)
 			}
-			authStm = fmt.Sprintf("%s AS 0x%s", authStm, hashed_hex)
+			authStm = fmt.Sprintf("%s AS 0x%s", authStm, hexDigits)
 		}
 
 	}
@@ -322,15 +321,14 @@ func UpdateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 			if d.Get("auth_string_hashed").(string) != "" {
 				authString = fmt.Sprintf("IDENTIFIED WITH %s AS '%s'", d.Get("auth_plugin"), d.Get("auth_string_hashed"))
 			} else if d.Get("auth_string_hex").(string) != "" {
-				hexValue := d.Get("auth_string_hex").(string)
-				if strings.HasPrefix(hexValue, "0x") || strings.HasPrefix(hexValue, "0X") {
-					hexValue = hexValue[2:]
-				}
-				if err := validateHexString(hexValue); err != nil {
+				authStringHex := d.Get("auth_string_hex").(string)
+				normalizedHex := normalizeHexString(authStringHex)
+
+				hexDigits := normalizedHex[2:]
+				if err := validateHexString(hexDigits); err != nil {
 					return diag.Errorf("invalid hex string for auth_string_hex: %v", err)
 				}
-
-				authString = fmt.Sprintf("IDENTIFIED WITH %s AS 0x%s", d.Get("auth_plugin"), hexValue)
+				authString = fmt.Sprintf("IDENTIFIED WITH %s AS 0x%s", d.Get("auth_plugin"), hexDigits)
 			}
 			stmtSQL = fmt.Sprintf("ALTER USER `%s`@`%s` %s  REQUIRE %s",
 				d.Get("user").(string),
@@ -489,9 +487,11 @@ func ReadUser(ctx context.Context, d *schema.ResourceData, meta interface{}) dia
 				}
 			} else {
 				quotedAuthString := m[4]
-				hexAuthString := m[5]
-				if hexAuthString != "" {
-					d.Set("auth_string_hex", hexAuthString)
+				authStringHex := m[5]
+
+				if authStringHex != "" {
+					normalizedHex := normalizeHexString(authStringHex)
+					d.Set("auth_string_hex", normalizedHex)
 					d.Set("auth_string_hashed", "")
 				} else if quotedAuthString != "" {
 					d.Set("auth_string_hashed", quotedAuthString)
@@ -582,6 +582,22 @@ func NewEmptyStringSuppressFunc(k, old, new string, d *schema.ResourceData) bool
 
 	return false
 }
+func SuppressHexStringDiff(k, old, new string, d *schema.ResourceData) bool {
+	if new == "" {
+		return true
+	}
+
+	// Normalize both values and compare
+	normalizedOld := normalizeHexString(old)
+	normalizedNew := normalizeHexString(new)
+
+	// Suppress diff if they're the same after normalization
+	if normalizedOld == normalizedNew {
+		return true
+	}
+
+	return false
+}
 
 func validateHexString(hexStr string) error {
 	if len(hexStr) == 0 {
@@ -599,4 +615,22 @@ func validateHexString(hexStr string) error {
 	}
 
 	return nil
+}
+
+// Add this helper function to normalize hex strings
+func normalizeHexString(hexStr string) string {
+	if hexStr == "" {
+		return ""
+	}
+
+	// Remove 0x prefix if present
+	if strings.HasPrefix(hexStr, "0x") || strings.HasPrefix(hexStr, "0X") {
+		hexStr = hexStr[2:]
+	}
+
+	// Convert to lowercase for consistency
+	hexStr = strings.ToUpper(hexStr)
+
+	// Always return with 0x prefix for consistency
+	return "0x" + hexStr
 }

--- a/mysql/resource_user.go
+++ b/mysql/resource_user.go
@@ -315,14 +315,24 @@ func UpdateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		auth = v.(string)
 	}
 	if len(auth) > 0 {
-		if d.HasChange("tls_option") || d.HasChange("auth_plugin") || d.HasChange("auth_string_hashed") {
+		if d.HasChange("tls_option") || d.HasChange("auth_plugin") || d.HasChange("auth_string_hashed") || d.HasChange("auth_string_hex") {
 			var stmtSQL string
 
 			authString := ""
 			if d.Get("auth_string_hashed").(string) != "" {
 				authString = fmt.Sprintf("IDENTIFIED WITH %s AS '%s'", d.Get("auth_plugin"), d.Get("auth_string_hashed"))
+			} else if d.Get("auth_string_hex").(string) != "" {
+				hexValue := d.Get("auth_string_hex").(string)
+				if strings.HasPrefix(hexValue, "0x") || strings.HasPrefix(hexValue, "0X") {
+					hexValue = hexValue[2:]
+				}
+				if err := validateHexString(hexValue); err != nil {
+					return diag.Errorf("invalid hex string for auth_string_hex: %v", err)
+				}
+
+				authString = fmt.Sprintf("IDENTIFIED WITH %s AS 0x%s", d.Get("auth_plugin"), hexValue)
 			}
-			stmtSQL = fmt.Sprintf("ALTER USER '%s'@'%s' %s  REQUIRE %s",
+			stmtSQL = fmt.Sprintf("ALTER USER `%s`@`%s` %s  REQUIRE %s",
 				d.Get("user").(string),
 				d.Get("host").(string),
 				authString,
@@ -414,10 +424,15 @@ func ReadUser(ctx context.Context, d *schema.ResourceData, meta interface{}) dia
 	}
 	requiredVersion, _ := version.NewVersion("5.7.0")
 	if getVersionFromMeta(ctx, meta).GreaterThan(requiredVersion) {
-		stmt := "SHOW CREATE USER ?@?"
 
+		_, err := db.ExecContext(ctx, "SET print_identified_with_as_hex = ON")
+		if err != nil {
+			// return diag.Errorf("failed setting print_identified_with_as_hex: %v", err)
+			log.Printf("[DEBUG] Could not set print_identified_with_as_hex: %v", err)
+		}
+		stmt := "SHOW CREATE USER ?@?"
 		var createUserStmt string
-		err := db.QueryRowContext(ctx, stmt, d.Get("user").(string), d.Get("host").(string)).Scan(&createUserStmt)
+		err = db.QueryRowContext(ctx, stmt, d.Get("user").(string), d.Get("host").(string)).Scan(&createUserStmt)
 		if err != nil {
 			errorNumber := mysqlErrorNumber(err)
 			if errorNumber == unknownUserErrCode || errorNumber == userNotFoundErrCode {
@@ -426,17 +441,17 @@ func ReadUser(ctx context.Context, d *schema.ResourceData, meta interface{}) dia
 			}
 			return diag.Errorf("failed getting user: %v", err)
 		}
-
 		// Examples of create user:
 		// CREATE USER 'some_app'@'%' IDENTIFIED WITH 'mysql_native_password' AS '*0something' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK
 		// CREATE USER `jdoe-tf-test-47`@`example.com` IDENTIFIED WITH 'caching_sha2_password' REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK PASSWORD HISTORY DEFAULT PASSWORD REUSE INTERVAL DEFAULT PASSWORD REQUIRE CURRENT DEFAULT
 		// CREATE USER `jdoe`@`example.com` IDENTIFIED WITH 'caching_sha2_password' AS '$A$005$i`xay#fG/\' TrbkNA82' REQUIRE NONE PASSWORD
-		re := regexp.MustCompile("^CREATE USER ['`]([^'`]*)['`]@['`]([^'`]*)['`] IDENTIFIED WITH ['`]([^'`]*)['`] (?:AS '((?:.*?[^\\\\])?)' )?REQUIRE ([^ ]*)")
-		if m := re.FindStringSubmatch(createUserStmt); len(m) == 6 {
+		// CREATE USER `hashed_hex`@`localhost` IDENTIFIED WITH 'caching_sha2_password' AS 0x244124303035242522434C16580334755221766C29210D2C415E033550367655494F314864686775414E735A742E6F474857504B623172525066574D524F30506B7A79646F30 REQUIRE NONE PASSWORD EXPIRE DEFAULT ACCOUNT UNLOCK PASSWORD HISTORY DEFAULT PASSWORD REUSE INTERVAL DEFAULT PASSWORD REQUIRE CURRENT DEFAULT
+		re := regexp.MustCompile("^CREATE USER ['`]([^'`]*)['`]@['`]([^'`]*)['`] IDENTIFIED WITH ['`]([^'`]*)['`] (?:AS (?:'((?:.*?[^\\\\])?)'|(0x[0-9A-Fa-f]+)) )?REQUIRE ([^ ]*)")
+		if m := re.FindStringSubmatch(createUserStmt); len(m) == 7 {
 			d.Set("user", m[1])
 			d.Set("host", m[2])
 			d.Set("auth_plugin", m[3])
-			d.Set("tls_option", m[5])
+			d.Set("tls_option", m[6])
 
 			if m[3] == "aad_auth" {
 				// AADGroup:98e61c8d-e104-4f8c-b1a6-7ae873617fe6:upn:Doe_Family_Group
@@ -473,7 +488,18 @@ func ReadUser(ctx context.Context, d *schema.ResourceData, meta interface{}) dia
 					return diag.Errorf("AAD identity couldn't be parsed - it is %s", m[4])
 				}
 			} else {
-				d.Set("auth_string_hashed", m[4])
+				quotedAuthString := m[4]
+				hexAuthString := m[5]
+				if hexAuthString != "" {
+					d.Set("auth_string_hex", hexAuthString)
+					d.Set("auth_string_hashed", "")
+				} else if quotedAuthString != "" {
+					d.Set("auth_string_hashed", quotedAuthString)
+					d.Set("auth_string_hex", "")
+				} else {
+					d.Set("auth_string_hashed", "")
+					d.Set("auth_string_hex", "")
+				}
 			}
 			return nil
 		}

--- a/mysql/resource_user.go
+++ b/mysql/resource_user.go
@@ -95,7 +95,13 @@ func resourceUser() *schema.Resource {
 				DiffSuppressFunc: NewEmptyStringSuppressFunc,
 				ConflictsWith:    []string{"plaintext_password", "password"},
 			},
-
+			"auth_string_hex": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				Sensitive:        true,
+				DiffSuppressFunc: NewEmptyStringSuppressFunc,
+				ConflictsWith:    []string{"plaintext_password", "password", "auth_string_hashed"},
+			},
 			"tls_option": {
 				Type:     schema.TypeString,
 				Optional: true,
@@ -171,6 +177,29 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 			authStm = fmt.Sprintf("%s AS ?", authStm)
 		}
 	}
+	var hashed_hex string
+	if v, ok := d.GetOk("auth_string_hex"); ok {
+		hashed_hex = v.(string)
+		if hashed_hex != "" {
+			if hashed != "" {
+				return diag.Errorf("can not specify both auth_string_hashed and auth_string_hex")
+			}
+			if authStm == "" {
+				return diag.Errorf("auth_string_hex is not supported for auth plugin %s", auth)
+			}
+			if strings.HasPrefix(hashed_hex, "0x") || strings.HasPrefix(hashed_hex, "0X") {
+				hashed_hex = hashed_hex[2:]
+			}
+
+			if err := validateHexString(hashed_hex); err != nil {
+				return diag.Errorf("invalid hex string for auth_string_hex: %v", err)
+			}
+			authStm = fmt.Sprintf("%s AS 0x%s", authStm, hashed_hex)
+		}
+
+	}
+	user := d.Get("user").(string)
+	host := d.Get("host").(string)
 
 	var stmtSQL string
 	var args []interface{}
@@ -180,15 +209,15 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		if aadIdentity["type"].(string) == "service_principal" {
 			// CREATE AADUSER 'mysqlProtocolLoginName"@"mysqlHostRestriction' IDENTIFIED BY 'identityId'
 			stmtSQL = "CREATE AADUSER ?@? IDENTIFIED BY ?"
-			args = []interface{}{d.Get("user").(string), d.Get("host").(string), aadIdentity["identity"].(string)}
+			args = []interface{}{user, host, aadIdentity["identity"].(string)}
 		} else {
 			// CREATE AADUSER 'identityName"@"mysqlHostRestriction' AS 'mysqlProtocolLoginName'
 			stmtSQL = "CREATE AADUSER ?@? AS ?"
-			args = []interface{}{aadIdentity["identity"].(string), d.Get("host").(string), d.Get("user").(string)}
+			args = []interface{}{aadIdentity["identity"].(string), host, user}
 		}
 	} else {
 		stmtSQL = "CREATE USER ?@?"
-		args = []interface{}{d.Get("user").(string), d.Get("host").(string)}
+		args = []interface{}{user, host}
 	}
 
 	var password string
@@ -198,7 +227,7 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		password = d.Get("password").(string)
 	}
 
-	if auth == "AWSAuthenticationPlugin" && d.Get("host").(string) == "localhost" {
+	if auth == "AWSAuthenticationPlugin" && host == "localhost" {
 		return diag.Errorf("cannot use IAM auth against localhost")
 	}
 
@@ -223,7 +252,7 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 	if getVersionFromMeta(ctx, meta).GreaterThan(requiredVersion) && d.Get("tls_option").(string) != "" {
 		if createObj == "AADUSER" {
 			updateStmtSql = "ALTER USER ?@? REQUIRE " + d.Get("tls_option").(string)
-			updateArgs = []interface{}{d.Get("user").(string), d.Get("host").(string)}
+			updateArgs = []interface{}{user, host}
 		} else {
 			stmtSQL += " REQUIRE " + d.Get("tls_option").(string)
 		}
@@ -246,8 +275,8 @@ func CreateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) d
 		return diag.Errorf("failed executing SQL: %v", err)
 	}
 
-	user := fmt.Sprintf("%s@%s", d.Get("user").(string), d.Get("host").(string))
-	d.SetId(user)
+	userId := fmt.Sprintf("%s@%s", user, host)
+	d.SetId(userId)
 
 	if updateStmtSql != "" {
 		log.Println("[DEBUG] Executing statement:", updateStmtSql, "args:", updateArgs)
@@ -526,4 +555,22 @@ func NewEmptyStringSuppressFunc(k, old, new string, d *schema.ResourceData) bool
 	}
 
 	return false
+}
+
+func validateHexString(hexStr string) error {
+	if len(hexStr) == 0 {
+		return fmt.Errorf("hex string cannot be empty")
+	}
+
+	if len(hexStr)%2 != 0 {
+		return fmt.Errorf("hex string must have even length")
+	}
+
+	for i, char := range strings.ToLower(hexStr) {
+		if !((char >= '0' && char <= '9') || (char >= 'a' && char <= 'f')) {
+			return fmt.Errorf("invalid hex character '%c' at position %d", char, i)
+		}
+	}
+
+	return nil
 }

--- a/mysql/resource_user_test.go
+++ b/mysql/resource_user_test.go
@@ -119,22 +119,43 @@ func TestAccUser_auth_mysql8(t *testing.T) {
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
 				),
 			},
+		},
+	})
+}
+
+func TestAccUser_auth_string_hash_mysql8(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheckSkipTiDB(t)
+			testAccPreCheckSkipMariaDB(t)
+			testAccPreCheckSkipRds(t)
+			testAccPreCheckSkipNotMySQLVersionMin(t, "8.0.14")
+		},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccUserCheckDestroy,
+		Steps: []resource.TestStep{
 			{
 				Config: testAccUserConfig_auth_caching_sha2_password_hex_no_prefix,
 				Check: resource.ComposeTestCheckFunc(
 					testAccUserAuthExists("mysql_user.test"),
-					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
-					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "hex"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "%"),
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
-					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"),
+				),
+			},
+			{
+				Config: testAccUserConfig_auth_caching_sha2_password_hex_no_prefix,
+				Check: resource.ComposeTestCheckFunc(
+					testAccUserAuthValid("hex", "password"),
 				),
 			},
 			{
 				Config: testAccUserConfig_auth_caching_sha2_password_hex_with_prefix,
 				Check: resource.ComposeTestCheckFunc(
 					testAccUserAuthExists("mysql_user.test"),
-					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
-					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "hex"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "%"),
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"),
 				),
@@ -143,8 +164,8 @@ func TestAccUser_auth_mysql8(t *testing.T) {
 				Config: testAccUserConfig_auth_caching_sha2_password_hex_updated,
 				Check: resource.ComposeTestCheckFunc(
 					testAccUserAuthExists("mysql_user.test"),
-					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
-					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "hex"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "%"),
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"),
 				),
@@ -585,24 +606,24 @@ resource "mysql_user" "test" {
 
 const testAccUserConfig_auth_caching_sha2_password_hex_no_prefix = `
 resource "mysql_user" "test" {
-    user            = "jdoe"
-    host            = "example.com"
+    user            = "hex"
+    host            = "%"
     auth_plugin     = "caching_sha2_password"
-    auth_string_hex = "244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"
+    auth_string_hex = "244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"
 }
 `
 const testAccUserConfig_auth_caching_sha2_password_hex_with_prefix = `
 resource "mysql_user" "test" {
-    user            = "jdoe"
-    host            = "example.com"
+    user            = "hex"
+    host            = "%"
     auth_plugin     = "caching_sha2_password"
     auth_string_hex = "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"
 }
 `
 const testAccUserConfig_auth_caching_sha2_password_hex_updated = `
 resource "mysql_user" "test" {
-    user            = "jdoe"
-    host            = "example.com"
+    user            = "hex"
+    host            = "%"
     auth_plugin     = "caching_sha2_password"
     auth_string_hex = "244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"
 }

--- a/mysql/resource_user_test.go
+++ b/mysql/resource_user_test.go
@@ -174,7 +174,7 @@ func TestAccUser_auth_mysql8_validation(t *testing.T) {
 			},
 			{
 				Config:      testAccUserConfig_auth_both_string_fields,
-				ExpectError: regexp.MustCompile(`can not specify both auth_string_hashed and auth_string_hex`),
+				ExpectError: regexp.MustCompile(`"auth_string_hex": conflicts with auth_string_hashed`),
 			},
 		},
 	})

--- a/mysql/resource_user_test.go
+++ b/mysql/resource_user_test.go
@@ -119,6 +119,63 @@ func TestAccUser_auth_mysql8(t *testing.T) {
 					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
 				),
 			},
+			{
+				Config: testAccUserConfig_auth_caching_sha2_password_hex_no_prefix,
+				Check: resource.ComposeTestCheckFunc(
+					testAccUserAuthExists("mysql_user.test"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"),
+				),
+			},
+			{
+				Config: testAccUserConfig_auth_caching_sha2_password_hex_with_prefix,
+				Check: resource.ComposeTestCheckFunc(
+					testAccUserAuthExists("mysql_user.test"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"),
+				),
+			},
+			{
+				Config: testAccUserConfig_auth_caching_sha2_password_hex_updated,
+				Check: resource.ComposeTestCheckFunc(
+					testAccUserAuthExists("mysql_user.test"),
+					resource.TestCheckResourceAttr("mysql_user.test", "user", "jdoe"),
+					resource.TestCheckResourceAttr("mysql_user.test", "host", "example.com"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_plugin", "caching_sha2_password"),
+					resource.TestCheckResourceAttr("mysql_user.test", "auth_string_hex", "0x244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccUser_auth_mysql8_validation(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheckSkipTiDB(t)
+			testAccPreCheckSkipMariaDB(t)
+			testAccPreCheckSkipRds(t)
+			testAccPreCheckSkipNotMySQLVersionMin(t, "8.0.14")
+		},
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccUserCheckDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccUserConfig_auth_caching_sha2_password_hex_invalid,
+				ExpectError: regexp.MustCompile(`invalid hex character 'g'`),
+			},
+			{
+				Config:      testAccUserConfig_auth_caching_sha2_password_hex_odd_length,
+				ExpectError: regexp.MustCompile(`hex string must have even length`),
+			},
+			{
+				Config:      testAccUserConfig_auth_both_string_fields,
+				ExpectError: regexp.MustCompile(`can not specify both auth_string_hashed and auth_string_hex`),
+			},
 		},
 	})
 }
@@ -523,6 +580,56 @@ resource "mysql_user" "test" {
     host               = "example.com"
     auth_plugin        = "caching_sha2_password"
     plaintext_password = "password"
+}
+`
+
+const testAccUserConfig_auth_caching_sha2_password_hex_no_prefix = `
+resource "mysql_user" "test" {
+    user            = "jdoe"
+    host            = "example.com"
+    auth_plugin     = "caching_sha2_password"
+    auth_string_hex = "244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"
+}
+`
+const testAccUserConfig_auth_caching_sha2_password_hex_with_prefix = `
+resource "mysql_user" "test" {
+    user            = "jdoe"
+    host            = "example.com"
+    auth_plugin     = "caching_sha2_password"
+    auth_string_hex = "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756576B317A5064687A715347765747746B66746A5A4F6E384C41756E6750495330"
+}
+`
+const testAccUserConfig_auth_caching_sha2_password_hex_updated = `
+resource "mysql_user" "test" {
+    user            = "jdoe"
+    host            = "example.com"
+    auth_plugin     = "caching_sha2_password"
+    auth_string_hex = "244124303035242931790D223576077A1446190832544A61301A256D5245316662534E56317A434A6A625139555A5642486F4B7A6F675266656B583330744379783134313239"
+}
+`
+const testAccUserConfig_auth_caching_sha2_password_hex_invalid = `
+resource "mysql_user" "test" {
+    user            = "jdoe"
+    host            = "example.com"
+    auth_plugin     = "caching_sha2_password"
+    auth_string_hex = "0x244124303035246g4f1e0d5d1631594f5c56701f3d327d073a724c706273307a5965516c7756"
+}
+`
+const testAccUserConfig_auth_caching_sha2_password_hex_odd_length = `
+resource "mysql_user" "test" {
+    user            = "jdoe"
+    host            = "example.com"
+    auth_plugin     = "caching_sha2_password"
+    auth_string_hex = "0x244124303035246c4f1e0d5d1631594f5c56701f3d327d073a724c706273307a5965516c775"
+}
+`
+const testAccUserConfig_auth_both_string_fields = `
+resource "mysql_user" "test" {
+    user                = "jdoe"
+    host                = "example.com"
+    auth_plugin         = "caching_sha2_password"
+    auth_string_hashed  = "*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19"
+    auth_string_hex     = "0x244124303035246c4f1e0d5d1631594f5c56701f3d327d073a724c706273307a5965516c7756"
 }
 `
 

--- a/website/docs/r/user.html.markdown
+++ b/website/docs/r/user.html.markdown
@@ -48,13 +48,22 @@ resource "mysql_user" "nologin" {
 ```
 
 ## Example Usage with caching_sha2_password Authentication Plugin and plaintext password
-
 ```hcl
 resource "mysql_user" "nologin" {
   user               = "nologin"
   host               = "example.com"
   auth_plugin        = "caching_sha2_password"
   plaintext_password = "password"
+}
+```
+## Example Usage with caching_sha2_password Authentication with Hex Hash
+
+```hcl
+resource "mysql_user" "nologin" {
+  user            = "nologin"
+  host            = "example.com"
+  auth_plugin     = "caching_sha2_password"
+  auth_string_hex = "0x244124303035246C4F1E0D5D1631594F5C56701F3D327D073A724C706273307A5965516C7756"
 }
 ```
 
@@ -83,6 +92,7 @@ The following arguments are supported:
 * `password` - (Optional) Deprecated alias of `plaintext_password`, whose value is _stored as plaintext in state_. Prefer to use `plaintext_password` instead, which stores the password as an unsalted hash.
 * `auth_plugin` - (Optional) Use an [authentication plugin][ref-auth-plugins] to authenticate the user instead of using password authentication.  Description of the fields allowed in the block below.
 * `auth_string_hashed` - (Optional) Use an already hashed string as a parameter to `auth_plugin`. This can be used with passwords as well as with other auth strings.
+* `auth_string_hex` - (Optional) The authentication string as a hexadecimal value(can be with or without `0x` prefix). Primarily used with `caching_sha2_password` authentication plugin. Cannot be used with `plaintext_password`, `password`, or `auth_string_hashed`.
 * `aad_identity` - (Optional) Required when `auth_plugin` is `aad_auth`. This should be block containing `type` and `identity`. `type` can be one of `user`, `group` and `service_principal`. `identity` then should containt either UPN of user, name of group or Client ID of service principal.
 * `retain_old_password` - (Optional) When `true`, the old password is retained when changing the password. Defaults to `false`. This use MySQL Dual Password Support feature and requires MySQL version 8.0.14 or newer. See [MySQL Dual Password documentation](https://dev.mysql.com/doc/refman/8.0/en/password-management.html#dual-passwords) for more.
 * `discard_old_password` - (Optional) When `true`, the old password is deleted. Defaults to `false`. This use MySQL Dual Password Support feature and requires MySQL version 8.0.14 or newer. See [MySQL Dual Password documentation](https://dev.mysql.com/doc/refman/8.0/en/password-management.html#dual-passwords) for more.

